File: lib/llm/core/agent.ts
import { LLM, CONTEXT } from '@/lib/constants';
import { Action } from './agentActions';
import { AgentErrorType } from './agentError';
import { getProjectContextWithDirectoryStructureAndAnalysis } from '../context';
import {
  generateChangesSummary,
  fetchChatHistory,
  sendOperationUpdate,
  tryRevalidatePath,
} from './agentCommunication';
import { AgentError, classifyError, getErrorMessage } from './agentError';
import {
  executeActions,
  updateContext,
  updateContextWithTracking,
  executeReadActionsForContext,
} from './agentActions';
import { generateAndParseAgentResponse, forceExecutionMode } from './agentPromptParser';

type ActionExecutionResult = {
  success: boolean;
  error?: string;
  errorType?: AgentErrorType;
  errorDetails?: string;
  actions?: Action[];
};

/**
 * Agent class responsible for orchestrating project modifications and handling UI updates
 */
export class Agent {
  private projectId: number;
  private readonly MAX_ITERATIONS = 25; // Prevent infinite loops
  private readonly DEFAULT_TIMEOUT_MS = 90000; // 90 seconds timeout for AI completions
  private readonly DEFAULT_MAX_TOKENS = 60000; // Maximum tokens for AI completions

  constructor(projectId: number) {
    this.projectId = projectId;
    console.log(`🚀 Agent initialized for project ID: ${projectId}`);
  }

  /**
   * Main public method to run the agent to process a project modification request
   */
  async run(prompt: string): Promise<{
    success: boolean;
    error?: string;
    errorType?: string;
    errorDetails?: string;
  }> {
    console.log(`🤖 Processing modification request for project ID: ${this.projectId}`);
    const processingStart = Date.now();

    try {
      // Create timeout promise for safety
      const timeoutPromise = new Promise<never>((_, reject) => {
        setTimeout(() => {
          reject(
            new AgentError({
              type: 'timeout',
              message: 'Processing timeout exceeded',
            })
          );
        }, LLM.PROCESSING_TIMEOUT);
      });

      try {
        // Get context and run the agent to get actions
        let context = '';
        try {
          console.log(`🔍 Getting project context for projectId: ${this.projectId}`);
          context = await getProjectContextWithDirectoryStructureAndAnalysis(this.projectId, {
            maxSize: CONTEXT.MAX_CONTEXT_SIZE,
            excludeDirs: CONTEXT.EXCLUDE_DIRS,
          });
          console.log(`✅ Successfully retrieved project context`);
        } catch (contextError) {
          console.warn('⚠️ Error getting project context:', contextError);
          // Continue without context
        }

        // Run the agentic workflow with timeout protection
        console.log('⏳ Running agentic workflow with timeout race...');
        const pipelineResult = (await Promise.race([
          this.runAgentic(this.projectId, prompt, context),
          timeoutPromise,
        ])) as ActionExecutionResult;

        this.logPipelineResult(pipelineResult);

        // Execute actions if any were returned
        if (pipelineResult.actions && pipelineResult.actions.length > 0) {
          await executeActions(this.projectId, pipelineResult.actions, generateChangesSummary);
        } else {
          console.warn(
            `⚠️ No actions returned from pipeline. Pipeline result: ${JSON.stringify(pipelineResult)}`
          );

          // Send error with specific error type
          await sendOperationUpdate(
            this.projectId,
            'error',
            '',
            `I was unable to understand how to make the requested changes. Please try rephrasing your request.`,
            'error',
            pipelineResult.errorType || ('unknown' as AgentErrorType)
          );

          // Return error with type information
          return {
            success: false,
            error: pipelineResult.error || 'No actions returned from pipeline',
            errorType: pipelineResult.errorType || 'processing',
            errorDetails: pipelineResult.errorDetails,
          };
        }
      } catch (error) {
        // Handle error during processing
        console.error('❌ Error or timeout processing modification:', error);

        // Determine error type
        const errorType = classifyError(error);
        const errorMessage = getErrorMessage(error, errorType);

        await sendOperationUpdate(this.projectId, 'error', '', errorMessage, 'error', errorType);

        return {
          success: false,
          error: errorMessage,
          errorType,
          errorDetails: error instanceof Error ? error.stack : undefined,
        };
      }

      // Revalidate path if in web environment
      tryRevalidatePath(this.projectId);

      const processingEnd = Date.now();
      console.log(`⏱️ Total processing time: ${processingEnd - processingStart}ms`);

      return { success: true };
    } catch (error) {
      console.error(`❌ Error in Agent.run:`, error);

      // Classify the error
      const errorType = classifyError(error);
      const errorMessage = getErrorMessage(error, errorType);

      return {
        success: false,
        error: errorMessage,
        errorType,
        errorDetails: error instanceof Error ? error.stack : undefined,
      };
    }
  }

  /**
   * Run the agentic workflow where the model can iteratively read files and gather context
   *
   * This method orchestrates the agent's thinking process:
   * 1. Iteratively gathers context by reading files
   * 2. Tracks files that have been read to avoid duplicates
   * 3. Monitors iteration count to prevent infinite loops
   * 4. Handles the transition from thinking to execution mode
   *
   * @param projectId - The ID of the current project
   * @param prompt - The user's original request prompt
   * @param context - Initial context for the agent
   * @returns A result object with actions to execute or an error
   */
  private async runAgentic(
    projectId: number,
    prompt: string,
    context: string
  ): Promise<ActionExecutionResult> {
    console.log(`🔄 Running agentic workflow for project ID: ${projectId}`);

    const isThinking = true; // This is only used for the initial loop condition
    const executionLog: string[] = [];
    const gatheredContext: Record<string, string> = {};
    const readFiles = new Set<string>(); // Track unique files that have been read
    let iterationCount = 0;

    try {
      // Fetch chat history for context
      const chatHistory = await fetchChatHistory(projectId);
      console.log(`📝 Fetched ${chatHistory.length} previous chat messages for context`);

      // Initial context
      let currentContext = context;

      // Iterative loop for agentic behavior
      while (isThinking && iterationCount < this.MAX_ITERATIONS) {
        iterationCount++;
        console.log(`🔄 Starting iteration ${iterationCount} of agentic workflow`);

        try {
          // Send a "Thinking..." message before each AI completion call
          await sendOperationUpdate(this.projectId, 'read', '', 'Thinking...', 'pending');

          // Update context with read files tracking and iteration warnings
          currentContext = updateContextWithTracking(
            currentContext,
            readFiles,
            iterationCount,
            this.MAX_ITERATIONS
          );

          // Generate and parse AI response
          const actions = await generateAndParseAgentResponse(
            prompt,
            currentContext,
            chatHistory,
            this.DEFAULT_TIMEOUT_MS,
            this.DEFAULT_MAX_TOKENS
          );

          // Process the parsed response
          if (!actions.thinking) {
            // Agent is ready to execute changes
            console.log(
              `✅ Agent is ready to execute changes, found ${actions.actions.length} actions`
            );

            return {
              success: true,
              actions: actions.actions,
            };
          }

          // Check for duplicate read requests and potentially force execution mode
          if (await this.shouldForceExecution(actions.actions, readFiles, iterationCount)) {
            // Force execution mode with one final attempt
            const finalActions = await forceExecutionMode(
              prompt,
              currentContext,
              chatHistory,
              this.DEFAULT_TIMEOUT_MS,
              this.DEFAULT_MAX_TOKENS
            );

            return {
              success: true,
              actions: finalActions,
            };
          }

          // Execute read actions to gather more context
          await executeReadActionsForContext(
            actions.actions,
            projectId,
            executionLog,
            gatheredContext,
            readFiles
          );

          // Update context with gathered information
          currentContext = updateContext(currentContext, gatheredContext, executionLog);
        } catch (iterationError) {
          console.error(`❌ Error in iteration ${iterationCount}:`, iterationError);

          // Add error information to the context
          const errorMsg = `\n\n### ERROR IN PREVIOUS ITERATION:\n${iterationError}\n\nPlease try a different approach.\n`;
          const addSectionToContext = (await import('./agentPromptParser')).addSectionToContext;
          currentContext = addSectionToContext(
            currentContext,
            errorMsg,
            '### ERROR IN PREVIOUS ITERATION:'
          );

          // Continue to next iteration unless we're at the limit
          if (iterationCount >= this.MAX_ITERATIONS - 1) {
            throw iterationError;
          }
        }
      }

      if (iterationCount >= this.MAX_ITERATIONS) {
        console.warn(`⚠️ Reached maximum iterations (${this.MAX_ITERATIONS}) in agentic workflow`);
        return {
          success: false,
          error: `Reached maximum iterations (${this.MAX_ITERATIONS}) in agentic workflow`,
        };
      }

      // We shouldn't reach here if the workflow is properly structured
      return { success: false, error: 'Agentic workflow completed without actions' };
    } catch (error) {
      console.error(`❌ Error in runAgentic:`, error);
      return {
        success: false,
        error: `Error in agentic workflow: ${error instanceof Error ? error.message : String(error)}`,
      };
    }
  }

  /**
   * Determine if we should force the agent into execution mode based on read actions
   */
  private async shouldForceExecution(
    actions: Action[],
    readFiles: Set<string>,
    iterationCount: number
  ): Promise<boolean> {
    // Check for duplicate read requests
    const duplicateReads = actions
      .filter(a => a.action === 'readFile' && readFiles.has(a.filePath))
      .map(a => a.filePath);

    if (duplicateReads.length > 0) {
      console.warn(`⚠️ Agent is trying to reread files: ${duplicateReads.join(', ')}`);

      // Force execution mode if too many duplicates or too many iterations
      return duplicateReads.length >= 3 || iterationCount >= Math.floor(this.MAX_ITERATIONS * 0.8);
    }

    return false;
  }

  /**
   * Log details of the pipeline result
   */
  private logPipelineResult(pipelineResult: ActionExecutionResult) {
    console.log(`✅ Pipeline processing completed successfully`);
    console.log(`📊 Pipeline result success: ${pipelineResult.success}`);
    console.log(`📊 Pipeline result error: ${pipelineResult.error || 'none'}`);
    console.log(
      `📊 Pipeline result actions: ${pipelineResult.actions ? pipelineResult.actions.length : 0}`
    );
    if (pipelineResult.actions && pipelineResult.actions.length > 0) {
      console.log(`📊 First action: ${JSON.stringify(pipelineResult.actions[0])}`);
    }
    console.log(`📊 Full pipeline result: ${JSON.stringify(pipelineResult, null, 2)}`);
  }
}

File: agent/app/core/agent.py
import asyncio
from typing import Dict, List, Set, AsyncGenerator
from app.models.actions import Action, ActionExecutionResult
from app.models.exceptions import AgentError, AgentErrorType, classify_error, get_error_message
from app.models.requests import ChatMessage
from app.services.llm_service import llm_service
from app.services.fs_service import fs_service
from app.services.webhook_service import WebhookService
from app.core.actions import ActionExecutor
from app.utils.config import settings
from app.utils.token_counter import count_tokens
import time

class Agent:
    """
    Main Agent class responsible for orchestrating project modifications with streaming updates
    
    Mirrors the TypeScript Agent class from lib/llm/core/agent.ts
    """
    
    def __init__(self, project_id: int):
        self.project_id = project_id
        self.max_iterations = settings.max_iterations
        self.action_executor = ActionExecutor(project_id)
        self.webhook_service = WebhookService()
        self.start_time = time.time()
        self.total_actions = 0
        self.total_tokens = 0
        print(f"🚀 Agent initialized for project ID: {project_id}")
    
    async def run(self, prompt: str) -> AsyncGenerator[Dict, None]:
        """
        Main agent workflow with streaming updates
        
        Mirrors the TypeScript Agent.run method
        """
        print(f"🤖 Processing modification request for project ID: {self.project_id}")
        processing_start = time.time()
        
        try:
            # Get basic project context (simplified for now - will add full context service later)
            yield {"type": "thinking", "file_path": "", "message": "Analyzing project structure...", "status": "pending"}
            
            # Create a basic context for now
            context = await self._get_basic_context()
            
            # Run agentic workflow
            async for update in self._run_agentic_workflow(prompt, context):
                yield update
                
        except Exception as e:
            error_type = classify_error(e)
            yield {
                "type": "error",
                "file_path": "",
                "message": get_error_message(e, error_type),
                "status": "error",
                "error_type": error_type.value
            }
        
        processing_end = time.time()
        print(f"⏱️ Total processing time: {processing_end - processing_start:.2f}s")
    
    async def _get_basic_context(self) -> str:
        """Get basic project context (placeholder for full context service)"""
        try:
            project_path = fs_service.get_project_path(self.project_id)
            if not project_path.exists():
                return "Project directory not found."
            
            # Get basic file listing
            files = await fs_service.list_files_recursively(str(project_path))
            
            context = f"""
================================================================
Project Context
================================================================
Project ID: {self.project_id}
Project Path: {project_path}

Files ({len(files)} total):
{chr(10).join(files[:20])}  # Show first 20 files
{'...' if len(files) > 20 else ''}
================================================================
"""
            return context
        except Exception as e:
            print(f"Error getting basic context: {e}")
            return "Error loading project context"
    
    async def _run_agentic_workflow(
        self, 
        prompt: str, 
        context: str
    ) -> AsyncGenerator[Dict, None]:
        """
        Run the iterative agentic workflow
        
        Mirrors the TypeScript _runAgentic method from agent.ts
        """
        print(f"🔄 Running agentic workflow for project ID: {self.project_id}")
        
        iteration_count = 0
        read_files: Set[str] = set()
        current_context = context
        execution_log: List[str] = []
        gathered_context: Dict[str, str] = {}
        
        while iteration_count < self.max_iterations:
            iteration_count += 1
            
            yield {
                "type": "thinking", 
                "file_path": "",
                "message": f"Thinking... (iteration {iteration_count})", 
                "status": "pending"
            }
            
            try:
                # Update context with tracking info
                current_context = self._update_context_with_tracking(
                    current_context, read_files, iteration_count
                )
                
                # Generate AI response
                messages = self._build_messages(prompt, current_context, [])
                response = await llm_service.generate_completion(messages)
                
                # Parse response
                parsed = await llm_service.parse_agent_response(response)
                
                if not parsed["thinking"]:
                    # Agent is ready to execute
                    yield {"type": "thinking", "file_path": "", "message": "Ready to execute changes", "status": "completed"}
                    
                    # Execute actions
                    async for update in self._execute_actions(parsed["actions"]):
                        yield update
                    
                    # Send completion webhook
                    await self._send_completion_webhook(success=True)
                    
                    return
                
                # Check for duplicate reads and force execution if needed
                if self._should_force_execution(parsed["actions"], read_files, iteration_count):
                    yield {"type": "thinking", "file_path": "", "message": "Forcing execution mode", "status": "pending"}
                    
                    final_actions = await self._force_execution_mode(prompt, current_context)
                    async for update in self._execute_actions(final_actions):
                        yield update
                    
                    # Send completion webhook
                    await self._send_completion_webhook(success=True)
                    
                    return
                
                # Execute read actions
                async for update in self._execute_read_actions(
                    parsed["actions"], read_files, gathered_context, execution_log
                ):
                    yield update
                
                # Update context
                current_context = self._update_context(current_context, gathered_context, execution_log)
                
            except Exception as e:
                print(f"Error in iteration {iteration_count}: {e}")
                # Add error to context and continue
                error_context = f"\n\n### ERROR IN PREVIOUS ITERATION:\n{str(e)}\n\nPlease try a different approach.\n"
                current_context += error_context
        
        # Max iterations reached
        raise AgentError(
            AgentErrorType.PROCESSING,
            f"Reached maximum iterations ({self.max_iterations})",
            "The agent was unable to complete the task within the iteration limit"
        )
    
    async def _execute_actions(self, actions: List[Action]) -> AsyncGenerator[Dict, None]:
        """
        Execute a list of actions and stream updates
        
        Mirrors the TypeScript executeActions functionality
        """
        print(f"🔄 Found {len(actions)} actions to execute")
        
        for i, action in enumerate(actions):
            print(f"⏳ Executing action {i+1}/{len(actions)}: {action.action} on {action.file_path}")
            
            # Map action type to update type
            update_type = self._map_action_to_update_type(action.action.value)
            
            yield {
                "type": update_type,
                "file_path": action.file_path,
                "message": action.message,
                "status": "pending"
            }
            
            try:
                action_start = time.time()
                success = await self.action_executor.execute_action(action)
                action_end = time.time()
                
                print(f"{'✅' if success else '❌'} Action {i+1} execution {'succeeded' if success else 'failed'} in {action_end - action_start:.2f}s")
                
                if success:
                    # Send webhook for successful action
                    async with self.webhook_service as webhook:
                        await webhook.send_action(
                            project_id=self.project_id,
                            action_type=action.action.value,
                            path=action.file_path,
                            status="completed"
                        )
                    
                    self.total_actions += 1
                    
                    yield {
                        "type": update_type,
                        "file_path": action.file_path,
                        "message": action.message,
                        "status": "completed"
                    }
                else:
                    # Send webhook for failed action
                    async with self.webhook_service as webhook:
                        await webhook.send_action(
                            project_id=self.project_id,
                            action_type=action.action.value,
                            path=action.file_path,
                            status="error"
                        )
                    
                    yield {
                        "type": "error",
                        "file_path": action.file_path,
                        "message": f"Failed to {action.action} on {action.file_path}",
                        "status": "error"
                    }
                    return
                    
            except Exception as e:
                yield {
                    "type": "error",
                    "file_path": action.file_path,
                    "message": f"Error executing {action.action}: {str(e)}",
                    "status": "error"
                }
                return
        
        # Send completion message
        yield {
            "type": "completed",
            "file_path": "",
            "message": "All changes have been implemented successfully!",
            "status": "completed"
        }
    
    async def _execute_read_actions(
        self, 
        actions: List[Action], 
        read_files: Set[str],
        gathered_context: Dict[str, str],
        execution_log: List[str]
    ) -> AsyncGenerator[Dict, None]:
        """
        Execute read actions and gather context
        
        Mirrors the TypeScript executeReadActionsForContext function
        """
        read_actions = [a for a in actions if a.action.value == "readFile"]
        
        if not read_actions:
            print('No read actions to execute')
            return
        
        print(f"🧠 Agent is still in thinking mode, executing {len(read_actions)} read actions...")
        
        for action in read_actions:
            if action.file_path in read_files:
                print(f"⚠️ Skip reading already read file: {action.file_path}")
                continue
            
            read_files.add(action.file_path)
            execution_log.append(f"Read {action.file_path}")
            
            yield {
                "type": "read",
                "file_path": action.file_path,
                "message": action.message,
                "status": "pending"
            }
            
            try:
                project_path = fs_service.get_project_path(self.project_id)
                full_path = project_path / action.file_path
                
                content = await fs_service.read_file(str(full_path))
                
                # Count tokens for tracking
                file_tokens = count_tokens(content)
                print(f"📊 Read {action.file_path}: {file_tokens} tokens")
                
                gathered_context[action.file_path] = content
                
                yield {
                    "type": "read",
                    "file_path": action.file_path,
                    "message": f"Read {action.file_path} successfully ({file_tokens} tokens)",
                    "status": "completed"
                }
                
            except Exception as e:
                gathered_context[action.file_path] = f"Error: {str(e)}"
                yield {
                    "type": "error",
                    "file_path": action.file_path,
                    "message": f"Error reading {action.file_path}: {str(e)}",
                    "status": "error"
                }
    
    def _should_force_execution(self, actions: List[Action], read_files: Set[str], iteration_count: int) -> bool:
        """Determine if we should force execution mode"""
        duplicate_reads = [a for a in actions if a.action.value == "readFile" and a.file_path in read_files]
        
        return (
            len(duplicate_reads) >= 3 or 
            iteration_count >= int(self.max_iterations * 0.8)
        )
    
    async def _force_execution_mode(self, prompt: str, context: str) -> List[Action]:
        """Force the agent into execution mode"""
        print("⚠️ Forcing agent to execution mode due to duplicate reads or high iteration count")
        
        forced_context = context + "\n\n### SYSTEM NOTICE - FORCING EXECUTION MODE:\nYou've attempted to reread files multiple times or have used too many iterations. Based on the files you've already read, proceed to implementation immediately.\n"
        
        messages = self._build_messages(prompt, forced_context, [])
        response = await llm_service.generate_completion(messages)
        parsed = await llm_service.parse_agent_response(response)
        
        return parsed["actions"]
    
    def _build_messages(self, prompt: str, context: str, chat_history: List) -> List[ChatMessage]:
        """Build messages for LLM completion"""
        system_content = context if context else ""
        
        messages = [
            ChatMessage(role="system", content=system_content),
            ChatMessage(role="user", content=prompt.strip())
        ]
        
        return messages
    
    def _update_context_with_tracking(self, context: str, read_files: Set[str], iteration_count: int) -> str:
        """Update context with tracking information"""
        updated_context = context
        
        if read_files:
            files_section = f"\n\n### Already Read Files - DO NOT READ THESE AGAIN:\n"
            files_section += "\n".join(f"{i+1}. {file}" for i, file in enumerate(read_files))
            updated_context += files_section
        
        if iteration_count >= int(self.max_iterations * 0.6):
            warning = f"\n\n### WARNING - APPROACHING ITERATION LIMIT:\nYou have used {iteration_count} of {self.max_iterations} available iterations. Move to implementation phase soon to avoid termination.\n"
            updated_context += warning
        
        return updated_context
    
    def _update_context(self, context: str, gathered_context: Dict[str, str], execution_log: List[str]) -> str:
        """Update context with gathered information"""
        updated_context = context
        
        if gathered_context:
            updated_context += "\n\n### File Contents:\n\n"
            for file_path, content in gathered_context.items():
                updated_context += f"--- File: {file_path} ---\n{content}\n\n"
        
        if execution_log:
            updated_context += "\n\n### Execution Log:\n\n"
            for i, log in enumerate(execution_log):
                updated_context += f"{i+1}. {log}\n"
        
        return updated_context
    
    def _map_action_to_update_type(self, action: str) -> str:
        """Map action type to stream update type"""
        mapping = {
            "readFile": "read",
            "createFile": "create", 
            "editFile": "edit",
            "deleteFile": "delete",
            "createDirectory": "create",
            "removeDirectory": "delete",
            "search": "read"
        }
        return mapping.get(action, "unknown")
    
    async def _send_completion_webhook(self, success: bool = True):
        """Send completion webhook to Next.js"""
        try:
            duration = time.time() - self.start_time
            
            async with self.webhook_service as webhook:
                await webhook.send_completion(
                    project_id=self.project_id,
                    success=success,
                    total_actions=self.total_actions,
                    total_tokens=self.total_tokens,
                    duration=duration
                )
            
            print(f"✅ Sent completion webhook: {self.total_actions} actions, {duration:.2f}s")
        except Exception as e:
            print(f"❌ Failed to send completion webhook: {e}") 